net-libs/libbitcoinconsensus/libbitcoinconsensus-0.21.1.ebuild

```diff
@@ -0,0 +1,111 @@
+# Copyright 1999-2021 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+inherit autotools
+
+BITCOINCORE_COMMITHASH="194b9b8792d9b0798fdb570b79fa51f1d1f5ebaf"
+KNOTS_PV="${PV}.knots20210629"
+KNOTS_P="bitcoin-${KNOTS_PV}"
+
+DESCRIPTION="Bitcoin Core consensus library"
+HOMEPAGE="https://bitcoincore.org/ https://bitcoinknots.org/"
+SRC_URI="
+	https://github.com/bitcoin/bitcoin/archive/${BITCOINCORE_COMMITHASH}.tar.gz -> bitcoin-v${PV}.tar.gz
+	https://bitcoinknots.org/files/0.21.x/${KNOTS_PV}/${KNOTS_P}.patches.txz -> ${KNOTS_P}.patches.tar.xz
+"
+
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS="~amd64 ~arm ~arm64 ~mips ~ppc ~ppc64 ~x86 ~amd64-linux ~x86-linux"
+IUSE="+asm bitcoin_protocol_taproot knots"
+
+DEPEND="
+	>dev-libs/libsecp256k1-0.1_pre20200911:=[recovery,schnorr]
+"
+RDEPEND="${DEPEND}"
+BDEPEND="
+	>=sys-devel/autoconf-2.69
+	>=sys-devel/automake-1.13
+"
+
+DOCS=( doc/bips.md doc/release-notes.md doc/shared-libraries.md )
+
+S="${WORKDIR}/bitcoin-${BITCOINCORE_COMMITHASH}"
+
+pkg_pretend() {
+	if use knots; then
+		elog "You are building ${PN} from Bitcoin Knots."
+		elog "For more information, see:"
+		elog "https://bitcoinknots.org/files/0.21.x/${KNOTS_PV}/${KNOTS_P}.desc.html"
+	else
+		elog "You are building ${PN} from Bitcoin Core."
+		elog "For more information, see:"
+		elog "https://bitcoincore.org/en/2021/05/01/release-${PV}/"
+	fi
+	elog
+	ewarn "CAUTION: BITCOIN PROTOCOL CHANGE INCLUDED"
+	ewarn "This release adds enforcement of the Taproot protocol change to the Bitcoin"
+	ewarn "rules, beginning in November. Protocol changes require user consent to be"
+	ewarn "effective, and if enforced inconsistently within the community may compromise"
+	ewarn "your security or others! If you do not know what you are doing, learn more"
+	ewarn "before November. (You must make a decision either way - simply not upgrading"
+	ewarn "is insecure in all scenarios.)"
+	ewarn "To learn more, see https://bitcointaproot.cc"
+	if ! use bitcoin_protocol_taproot; then
```

Comment:
I don't think this is the right approach (but I _do_ agree with ensuring users are informed):

- You mentioned upstream doesn't do auto-upgrades, but does it warn at all on installation/upgrade 'manually'?

- (Additionally, I don't really want to go beyond what upstream did or didn't do. Gentoo isn't really the place to lobby or impose what you wish they'd done upstream. Not saying that's what happening here, I just want to be careful of the line, because I know it's tempting sometimes.)

- I'd much prefer a news item for people who have this set of packages installed or a pkg_postinst message. Note that in general, we assume users read elogs because they often
contain important information. (People won't be running the new binary until they explicitly restart e.g. bitcoind anyway).

- In terms of reasons against pkg_pretend dying, the main issue is that it hurts automation (the package manager can only detect failure after dependency resolution and so on), and we already have REQUIRED_USE for things like this.

A REQUIRED_USE setting which isn't satisfied by default would trigger a QA warning which we _could_ possibly waive, but I'm still more inclined to go for the news item or pkg_postinst here. Those are the two official methods for letting users know about upcoming changes.

What do you reckon?

(cc @candrews)

Reply:
FWIW, it appears to me that this is nothing that users should be notified *via* gentoo mechanisms. I know that this is probably a delicate topic for some (as it's bitcoin). But ultimately a use flag that does nothing besides preventing the package from being emerged if not set, is not something we should introduce nor encourage in ::gentoo.

Reply:
Upon reflection, I agree with @Flowdalic. Users should be subscribed to upstream's news notifications (mailing lists, whatever). We should really only be using Gentoo news items (and arguably pkg_postinst) for critical breaking changes.

This kind of thing is quite natural for Bitcoin. I think we should therefore be doing a pkg_postinst message _at most_.

Reply:
This *is* a critical breaking change, one that can potentially cost the user their life savings.

I brought the subject up at today's Bitcoin development meeting, and there didn't seem to be much agreement for foregoing explicitly user interaction.

Reply:
Then shouldn't **upstream** be doing something about it? Make bitcoind or whatever scream if a config option hasn't been set? Gentoo, being downstream, isn't the place to lobby for things you wish were done upstream?

There still is user interaction. People should not be doing unattended upgrades because people have to read pkg_postinst messages anyway (see e.g. PAM upgrades, OpenSSH, ...). If they are, they're going to have other problems.

Let's go a bit further too. I don't want to be the only people doing this. What's happening with Debian/Ubuntu .debs? Fedora/* RPMs?

Reply:
>Then shouldn't upstream be doing something about it? 

We don't have any automatic updating upstream, intentionally.

>Make bitcoind or whatever scream if a config option hasn't been set?

That was also discussed, but it won't be here for Taproot.

>I also don't see anything on https://bitcoin.org/en/download mentioning this.

It probably should. https://bitcoincore.org/ is pretty explicit about it.

> Can you walk me through what happens when I run Bitcoin Core on say, Windows, and there's an update available?

Nothing. You have to manually seek it out and choose to upgrade.

>People should not be doing unattended upgrades because people have to read pkg_postinst messages anyway (see e.g. PAM upgrades, OpenSSH, ...).

Telling the user after the fact is kinda too late?

>Let's go a bit further too, I don't want to be the only people doing this.. what's happening with Debian/Ubuntu .debs? Fedora/* RPMs?

We asked them not to package at all. I maintain the PPA, and plan to do something similar (debconf has a way to insist on explicit user interaction)

>Separately: I'm wondering why Bitcoin isn't available for download in the UK.

It is. What gave you the impression it isn't?

Reply:
> I brought the subject up at today's Bitcoin development meeting, and there didn't seem to be much agreement for foregoing explicitly user interaction.

I read said meeting and agree that this is caused by using bitcoin via a package manager that performs auto updates *and* by bitcoin core enabling new consensus rules on new updates by default. I really appreciate your effort here.

That said, I think we best we can do is educate Gentoo users about this situation, that every net-p2p/bitcoind update potentially comes with new defaults that they need to research about. But `if ! use new-consensus-rule; then die; fi` seems excessive, especially since it may aborts an upgrade process (if `--keep-going` is not enabled).

I am currently leaning towards a short warning message on every update, but in case of taproot, a news item seems also sensible. Other means could be discussed as well, like a guard file, that if not existent prevents bitcoind from starting.

Reply:
`REQUIRED_USE` seems to solve the problem of aborting other updates.

Another possibility is an additional dummy LICENSE which isn't part of the default-accepted lists?

Reply:
>>People should not be doing unattended upgrades because people have to read pkg_postinst messages anyway (see e.g. PAM upgrades, OpenSSH, ...).

>Telling the user after the fact is kinda too late?

But it's not after the fact. Users shouldn't be doing unattended upgrades. They will see the log and downgrade if necessary?

>>Separately: I'm wondering why Bitcoin isn't available for download in the UK.

>It is. What gave you the impression it isn't?

https://bitcoin.org/en/download says "(This software is presently not available for download in the UK, and download links will not work if you are located within the UK.)" and the links (as it says) don't work.

----
I'm okay with a one-off news item as @Flowdalic suggests to indicate to users they should _always_ check on major upgrades and we'll do an elog too. In future, just an elog. How about that?

Reply:
>But it's not after the fact. Users shouldn't be doing unattended upgrades. They will see the log and downgrade if necessary?

It is after the fact, by definition. PAM/OpenSSH notices are things to do after the update, not reasons one might not consent to the upgrade.

>https://bitcoin.org/en/download says "(This software is presently not available for download in the UK, and download links will not work if you are located within the UK.)" and the links (as it says) don't work.

This is bitcoin.org's owner playing political games. There isn't any actual problem.

>I'm okay with a one-off news item as @Flowdalic suggests to indicate to users they should always check on major upgrades and we'll do an elog too. In future, just an elog. How about that?

That doesn't ensure the user has consented. This also isn't generally recognisable as a major upgrade (0.21.0 -> 0.21.1).